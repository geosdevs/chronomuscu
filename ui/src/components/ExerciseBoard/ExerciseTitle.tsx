import { useContext, useRef, useState } from "react";
import { ExerciseBoardOnTitleEditContext } from "../../App";
import { ExerciseBoardData } from "../../app-types";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faHashtag } from "@fortawesome/free-solid-svg-icons";

type ExerciseTitleProps = {
  exerciseBoard: ExerciseBoardData;
};

export const MAX_LENGTH_EXERCISE_TITLE = 12;

export default function ExerciseTitle({ exerciseBoard }: ExerciseTitleProps) {
  const [isEditing, setIsEditing] = useState(false);
  const onTitleEdit = useContext(ExerciseBoardOnTitleEditContext);
  const inputTitleRef = useRef<null | HTMLInputElement>(null);

  function handleNewTitleSubmit(e: any) {
    if (
      typeof e.currentTarget?.value === "string" &&
      e.currentTarget?.value
    ) {
      onTitleEdit(e.currentTarget.value);
    }
    setIsEditing(false);
  }

  return (
    <div>
      {isEditing ? (
        <span className="text-2xl">
          <strong className="text-gunmetal">
            <FontAwesomeIcon icon={faHashtag} size="lg" />
            {exerciseBoard.id}
          </strong>{" "}
          <input
            list="exercise-title-datalist"
            type="text"
            id="exercise-title-input"
            className="bg-zinc-200 rounded-md max-w-32 md:max-w-none"
            onBlur={handleNewTitleSubmit}
            ref={inputTitleRef}
            onInput={(e) => {
              const titleLength = Number(e.currentTarget.value?.length) ?? 0;

              if (titleLength > MAX_LENGTH_EXERCISE_TITLE) {
                e.currentTarget.value = e.currentTarget.value.substring(
                  0,
                  MAX_LENGTH_EXERCISE_TITLE + 1
                );
              }
            }}
          ></input>
          <datalist id="exercise-title-datalist">
            <option>Push-up</option>
            <option>Pull-up</option>
            <option>Squat</option>
            <option>Deadlift</option>
          </datalist>
        </span>
      ) : (
        <h2 className="text-2xl hover:scale-105 duration-300">
          <strong className="text-gunmetal">
            <FontAwesomeIcon icon={faHashtag} size="lg" />
            {exerciseBoard.id}
          </strong>{" "}
          <span
            className="text-cordovan hover:text-ecru duration-300"
            onClick={(e) => {
              setIsEditing(true);

              if (inputTitleRef.current instanceof HTMLInputElement) {
                inputTitleRef.current.defaultValue =
                  e.currentTarget.innerText.trim();
              }
            }}
          >
            {exerciseBoard.exerciseName}
          </span>
        </h2>
      )}
    </div>
  );
}
