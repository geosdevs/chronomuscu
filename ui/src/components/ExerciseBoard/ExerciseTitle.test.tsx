import { ExerciseBoardData } from "../../app-types";
import ExerciseTitle from "./ExerciseTitle";
import renderer from "react-test-renderer";

describe("ExerciseTitle", () => {
  it("should render", () => {
    const boardData: ExerciseBoardData = {
      exerciseName: "Test exercise",
      id: 654,
    };
    const component = renderer.create(<ExerciseTitle exerciseBoard={boardData}/>);
    const jsonTree = component.toJSON();

    expect(jsonTree).toMatchSnapshot();
  });
});
