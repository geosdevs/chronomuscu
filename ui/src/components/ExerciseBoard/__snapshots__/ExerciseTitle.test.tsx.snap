// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExerciseTitle should render 1`] = `
<div>
  <h2
    className="text-2xl hover:scale-105 duration-300"
  >
    <strong
      className="text-gunmetal"
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-hashtag fa-lg "
        data-icon="hashtag"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M181.3 32.4c17.4 2.9 29.2 19.4 26.3 36.8L197.8 128h95.1l11.5-69.3c2.9-17.4 19.4-29.2 36.8-26.3s29.2 19.4 26.3 36.8L357.8 128H416c17.7 0 32 14.3 32 32s-14.3 32-32 32H347.1L325.8 320H384c17.7 0 32 14.3 32 32s-14.3 32-32 32H315.1l-11.5 69.3c-2.9 17.4-19.4 29.2-36.8 26.3s-29.2-19.4-26.3-36.8l9.8-58.7H155.1l-11.5 69.3c-2.9 17.4-19.4 29.2-36.8 26.3s-29.2-19.4-26.3-36.8L90.2 384H32c-17.7 0-32-14.3-32-32s14.3-32 32-32h68.9l21.3-128H64c-17.7 0-32-14.3-32-32s14.3-32 32-32h68.9l11.5-69.3c2.9-17.4 19.4-29.2 36.8-26.3zM187.1 192L165.8 320h95.1l21.3-128H187.1z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
      654
    </strong>
     
    <span
      className="text-cordovan hover:text-ecru duration-300"
      onClick={[Function]}
    >
      Test exercise
    </span>
  </h2>
</div>
`;
